utils/src/registry.rs

```diff
@@ -0,0 +1,306 @@
+            PackageRegistry::Warg => "io.warg",
+        }
+    }
+
+    fn profile_url(&self, pkg: &PackageSpec) -> String {
+        match self {
+            PackageRegistry::Wapm => {
+                format!(
+                    "https://wapm.io/{}/{}@{}",
+                    pkg.author, pkg.name, pkg.version
+                )
+            }
+            PackageRegistry::Warg => todo!(),
+        }
+    }
+
+    fn fq_name(&self, pkg: &PackageSpec) -> String {
+        match self {
+            PackageRegistry::Wapm => {
+                format!(
+                    "{}/{}/{}@{}:{}",
+                    self.namespace(),
+                    pkg.author,
+                    pkg.name,
+                    pkg.version,
+                    pkg.module
+                )
+            }
+            PackageRegistry::Warg => todo!(),
+        }
+    }
+
+    fn download_urls(&self, pkg: &PackageSpec) -> Vec<String> {
+        match self {
+            PackageRegistry::Wapm => {
+                vec![
+                    // For some very stupid reason, wasm binaries can sit in multiple locations. Hopefully this is the full list:
+                    format!("https://registry-cdn.wapm.io/contents/{}/{}/{}/{}.wasm", pkg.author, pkg.name, pkg.version, pkg.module),
+                    format!("https://registry-cdn.wapm.io/contents/{}/{}/{}/target/wasm32-wasi/release/{}.wasm", pkg.author, pkg.name, pkg.version, pkg.module)
+                ]
+            }
+            PackageRegistry::Warg => todo!(),
+        }
+    }
+    // even cooler....
+    //fn download(&self, pkg: &PackageSpec) -> Result<Bytes, ServalError> {
+    //    // do the work of downloading from this kind of registry
+    //}
+}
+
+/// Specification for a registry package
+#[derive(Debug, Clone, PartialEq)]
+pub struct PackageSpec {
+    pub registry: PackageRegistry,
+    pub author: String,
+    pub name: String,
+    pub version: String,
+    pub module: String,
+}
+
+impl PackageSpec {
+    // other useful functions here
+
+    pub fn profile_url(&self) -> String {
+        self.registry.profile_url(self)
+    }
+
+    pub fn download_urls(&self) -> Vec<String> {
+        self.registry.download_urls(self)
+    }
+
+    pub fn fq_name(&self) -> String {
+        self.registry.fq_name(self)
+    }
+
+    pub fn fq_digest(&self) -> String {
+        digest(self.fq_name())
+    }
+
+    pub fn namespace(&self) -> String {
+        format!("{}/{}", self.registry.namespace(), self.author)
+    }
+
+    pub fn binary_path(&self) -> PathBuf {
+        PathBuf::from(format!("/tmp/{}", self.fq_digest()))
+    }
+
+    pub fn is_binary_cached(&self) -> bool {
+        self.binary_path().exists()
+    }
+}
+
+/// Converts an identifier string to a `PackageSpec`
+impl TryFrom<std::string::String> for PackageSpec {
+    type Error = ServalError;
+    /**
+    This function matches a package specification string.
+    It supports a number of variants:
```

Comment:
I may be missing some variants... but at least they are all automatically tested ;)